tpm2/test/policy_test.go

```diff
@@ -981,3 +1006,156 @@ func TestPolicyAuthValue(t *testing.T) {
+
+	k, err := CreateLoaded{
+		ParentHandle: AuthHandle{
+			Handle: pk.Handle,
+			Name:   pk.Name,
+			Auth:   PasswordAuth(nil),
+		},
+		InPublic: New2BTemplate(&TPMTPublic{
+			Type:    TPMAlgRSA,
+			NameAlg: TPMAlgSHA256,
+			ObjectAttributes: TPMAObject{
+				FixedTPM:            false,
+				FixedParent:         false,
+				SensitiveDataOrigin: true,
+				UserWithAuth:        true,
+				SignEncrypt:         true,
+			},
+			Parameters: NewTPMUPublicParms(
+				TPMAlgRSA,
+				&TPMSRSAParms{
+					Scheme: TPMTRSAScheme{
+						Scheme: TPMAlgRSASSA,
+						Details: NewTPMUAsymScheme(
+							TPMAlgRSASSA,
+							&TPMSSigSchemeRSASSA{
+								HashAlg: TPMAlgSHA256,
+							},
+						),
+					},
+					KeyBits: 2048,
+				},
+			),
+		}),
+	}.Execute(thetpm)
+	if err != nil {
+		t.Fatalf("error creating key %v", err)
+	}
+	defer func() {
+		t.Helper()
+		_, err := FlushContext{
+			FlushHandle: k.ObjectHandle,
+		}.Execute(thetpm)
+		if err != nil {
+			t.Errorf("error cleaning up key: %v", err)
+		}
+	}()
+
+	// create a trial policy with PolicyDuplicationSelect
+	sess, cleanup1, err := PolicySession(thetpm, TPMAlgSHA256, 16, Trial())
+	if err != nil {
+		t.Fatalf("error setting up trial session: %v", err)
+	}
+	defer func() {
+		t.Helper()
+		if err := cleanup1(); err != nil {
+			t.Errorf("error cleaning up trial session: %v", err)
+		}
+	}()
+
+	tests := []struct {
+		name          string
+		IncludeObject bool
+	}{
+		{"IncludeObjectFalse", false},
+		{"IncludeObjectTrue", true},
+	}
+
+	var pds PolicyDuplicationSelect
+	var pdr *PolicyGetDigestResponse
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if tt.IncludeObject {
+				pds = PolicyDuplicationSelect{
+					PolicySession: sess.Handle(),
+					NewParentName: ek.Name,
+					ObjectName:    k.Name,
+					IncludeObject: tt.IncludeObject,
+				}
+				_, err = pds.Execute(thetpm)
+				if err != nil {
+					t.Fatalf("error executing PolicyDuplicationSelect: %v", err)
+				}
+
+				pdr, err = PolicyGetDigest{
+					PolicySession: sess.Handle(),
+				}.Execute(thetpm)
+				if err != nil {
+					t.Fatalf("error executing PolicyGetDigest: %v", pdr)
+				}
+
+			} else {
+				// create a trial policy with PolicyDuplicationSelect
+				sess, cleanup, err := PolicySession(thetpm, TPMAlgSHA256, 16, Trial())
+				if err != nil {
+					t.Fatalf("error setting up trial session: %v", err)
+				}
+				defer func() {
+					t.Helper()
+					if err := cleanup(); err != nil {
+						t.Errorf("error cleaning up trial session: %v", err)
+					}
+				}()
+
+				pds = PolicyDuplicationSelect{
+					PolicySession: sess.Handle(),
+					NewParentName: ek.Name,
+				}
+				_, err = pds.Execute(thetpm)
+				if err != nil {
+					t.Fatalf("error executing PolicyDuplicationSelect: %v", err)
+				}
+
+				pdr, err = PolicyGetDigest{
+					PolicySession: sess.Handle(),
+				}.Execute(thetpm)
+				if err != nil {
+					t.Fatalf("error executing PolicyGetDigest: %v", pdr)
+				}
+
+			}
```

Comment:
I still don't quite follow why we need a lot of duplicated code between `if` and `else` here (or even, why we need an if statement at all). What I meant to suggest in https://github.com/google/go-tpm/pull/365/files#r1704689386 is the sequence:

* Create trial session
* Create PolicyDuplicationSelect object where `IncludeObject` is equal to `tt.IncludeObject`
* Run that policy and get the hash
* Create a policy calculator
* Feed the PolicyDuplicationSelect object created above into the calculator
* Check that the hash is the same

Reply:
the reason for the if statements is i wanted to explictly omit the null value ` TPM2BName{}` when sending in the `objectname` and includeobject values...thats more natural use to me.  

this is a testcase so not a big deal so added in this

```golang
	tests := []struct {
		name          string
		objectName    TPM2BName
		includeObject bool
	}{
		{"IncludeObjectFalse", TPM2BName{}, false},
		{"IncludeObjectTrue", k.Name, true},
	}

	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			pds := PolicyDuplicationSelect{
				PolicySession: sess.Handle(),
				NewParentName: ek.Name,
				ObjectName:    tt.objectName,
				IncludeObject: tt.includeObject,
			}
```

Reply:
Yeah, not a big deal either way but thank you for making this test case more concise!